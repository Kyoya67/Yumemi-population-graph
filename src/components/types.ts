export interface Prefecture {
    prefCode: number;
    prefName: string;
}

export interface PopulationData {
    prefName: string;
    data: {
        year: number;
        value: number;
    }[];
}

export type PopulationType = '総人口' | '年少人口' | '生産年齢人口' | '老年人口';
