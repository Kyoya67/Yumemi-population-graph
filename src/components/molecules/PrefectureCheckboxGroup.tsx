import { Checkbox } from '../atoms/Checkbox';
import { Prefecture } from '../types';

interface PrefectureCheckboxGroupProps {
    prefectures: Prefecture[];
    onPrefectureChange: (prefecture_code: number, prefecture_name: string, checked: boolean) => void;
    disabled?: boolean;
}

export const PrefectureCheckboxGroup = ({
    prefectures,
    onPrefectureChange,
    disabled
}: PrefectureCheckboxGroupProps) => (
    <div style={{ display: 'flex', flexWrap: 'wrap', gap: '10px', marginBottom: '20px' }}>
        {prefectures.map((pref) => (
            <Checkbox
                key={pref.prefCode}
                onChange={(checked) => onPrefectureChange(pref.prefCode, pref.prefName, checked)}
                disabled={disabled}
                label={pref.prefName}
            />
        ))}
    </div>
);