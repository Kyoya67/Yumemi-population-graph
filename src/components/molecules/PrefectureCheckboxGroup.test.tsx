import React from 'react';
import '@testing-library/jest-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import { PrefectureCheckboxGroup } from './PrefectureCheckboxGroup';
import { Prefecture } from '@/components/types';

describe('PrefectureCheckboxGroup', () => {
    const mockPrefectures: Prefecture[] = [
        { prefCode: 1, prefName: '北海道' },
        { prefCode: 13, prefName: '東京都' },
        { prefCode: 27, prefName: '大阪府' }
    ];

    const mockOnPrefectureChange = jest.fn();

    beforeEach(() => {
        mockOnPrefectureChange.mockClear();
    });

    describe('Rendering', () => {
        it('renders all prefecture checkboxes', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            mockPrefectures.forEach(pref => {
                const checkbox = screen.getByLabelText(pref.prefName);
                expect(checkbox).toBeInTheDocument();
                expect(checkbox).not.toBeChecked();
            });
        });

        it('renders with correct HTML attributes', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            mockPrefectures.forEach(pref => {
                const checkbox = screen.getByLabelText(pref.prefName);
                expect(checkbox).toHaveAttribute('type', 'checkbox');
            });
        });

        it('renders empty state when no prefectures provided', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={[]}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            expect(screen.queryByRole('checkbox')).not.toBeInTheDocument();
        });
    });

    describe('Interactions', () => {
        it('calls onPrefectureChange when checkbox is clicked', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            const checkbox = screen.getByLabelText('北海道');
            fireEvent.click(checkbox);
            expect(mockOnPrefectureChange).toHaveBeenCalledWith(1, '北海道', true);

            fireEvent.click(checkbox);
            expect(mockOnPrefectureChange).toHaveBeenCalledWith(1, '北海道', false);
        });

        it('handles multiple checkbox selections and deselections', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            const hokkaido = screen.getByLabelText('北海道');
            const tokyo = screen.getByLabelText('東京都');

            fireEvent.click(hokkaido);
            fireEvent.click(tokyo);
            fireEvent.click(hokkaido);

            expect(mockOnPrefectureChange).toHaveBeenCalledTimes(3);
            expect(mockOnPrefectureChange).toHaveBeenNthCalledWith(1, 1, '北海道', true);
            expect(mockOnPrefectureChange).toHaveBeenNthCalledWith(2, 13, '東京都', true);
            expect(mockOnPrefectureChange).toHaveBeenNthCalledWith(3, 1, '北海道', false);
        });
    });

    describe('Disabled state', () => {
        it('disables all checkboxes when disabled prop is true', () => {
            render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={true}
                />
            );

            mockPrefectures.forEach(pref => {
                const checkbox = screen.getByLabelText(pref.prefName);
                expect(checkbox).toBeDisabled();

                fireEvent.click(checkbox);
                expect(mockOnPrefectureChange).not.toHaveBeenCalled();
            });
        });

        it('enables checkboxes when disabled prop changes from true to false', () => {
            const { rerender } = render(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={true}
                />
            );

            rerender(
                <PrefectureCheckboxGroup
                    prefectures={mockPrefectures}
                    onPrefectureChange={mockOnPrefectureChange}
                    disabled={false}
                />
            );

            mockPrefectures.forEach(pref => {
                expect(screen.getByLabelText(pref.prefName)).not.toBeDisabled();
            });
        });
    });

    it('matches snapshot', () => {
        const { container } = render(
            <PrefectureCheckboxGroup
                prefectures={mockPrefectures}
                onPrefectureChange={mockOnPrefectureChange}
                disabled={false}
            />
        );
        expect(container).toMatchSnapshot();
    });
});