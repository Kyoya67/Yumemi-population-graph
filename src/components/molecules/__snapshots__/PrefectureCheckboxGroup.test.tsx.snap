// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrefectureCheckboxGroup matches snapshot 1`] = `
<div>
  <div
    class="flex flex-wrap gap-2 mb-5"
  >
    <label
      class="flex items-center p-1 cursor-pointer"
    >
      <input
        class="cursor-inherit"
        type="checkbox"
      />
      <span
        class="ml-1"
      >
        北海道
      </span>
    </label>
    <label
      class="flex items-center p-1 cursor-pointer"
    >
      <input
        class="cursor-inherit"
        type="checkbox"
      />
      <span
        class="ml-1"
      >
        東京都
      </span>
    </label>
    <label
      class="flex items-center p-1 cursor-pointer"
    >
      <input
        class="cursor-inherit"
        type="checkbox"
      />
      <span
        class="ml-1"
      >
        大阪府
      </span>
    </label>
  </div>
</div>
`;
